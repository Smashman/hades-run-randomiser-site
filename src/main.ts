import { Run, RunOptions } from './run';

const runOptions: RunOptions = {
    weaponOptions: {
        enabled: true,
        randomAspect: true,
    },
    keepsakeOptions: {
        enabled: true,
        ignoreMaxLevel: true,
        changeBetweenBiomes: true,
    },
    companionOptions: {
        enabled: true,
        ignoreMaxCodex: true,
    },
    mirrorOptions: {
        enabled: true,
    },
    pactOptions: {
        enabled: true,
        heatRange: {
            min: 4,
            max: 12,
        },
    }
};

const run = Run.generateRandomRun(runOptions);

console.log('run', run);
console.log('run.weapon', run.weapon);
console.log('run.weaponAspect', run.weaponAspect);
console.log('run.keepsake', run.keepsake);
console.log('run.companion', run.companion);
console.log('run.mirror', run.mirror);
console.log('run.pact', run.pact);