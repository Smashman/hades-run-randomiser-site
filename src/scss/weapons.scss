@import './style.scss';

.weapons {
    display: flex;
    flex-wrap: wrap;
}

.weapon, .aspect {
    .weapon-name, .aspect-name {
        font-family: $spectral-sc;
        font-weight: bold;
    }
    .weapon-title {
        font-family: $crimson-text;
        font-style: italic;
    }
}

.weapon {
    border: 1px black solid;
    width: 300px;
    .weapon-name {
        font-size: 1.3em;
    }
}

.aspect {
    border: 1px blue solid;
    height: 90px;
    .aspect-name {
        font-family: $alegreya-sans-sc;
    }
}