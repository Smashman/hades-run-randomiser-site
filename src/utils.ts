export const getRandomItemFromArray = <T>(array: T[]) => array[Math.floor(Math.random() * array.length)];
export const fillArray = <T>(length: number, fillValue: T) => Array.from<T>({length: length}).fill(fillValue);

export interface Options {
    enabled?: boolean;
}

export abstract class Item {
    isUnlocked: boolean = false;

    unlock() {
        this.isUnlocked = true;
    }

    lock() {
        this.isUnlocked = false;
    }
}

export abstract class ItemList<T extends Item> {
    items: T[] = [];

    addItem(item: T) {
        this.items = [...this.items, item];
        return this;
    }

    addItems(items: T[]) {
        items.forEach(item => this.addItem(item));
        return this;
    }

    get unlocked() {
        return this.items.filter(item => item.isUnlocked);
    }

    unlockAll() {
        this.items.forEach(item => item.unlock());
    }

    getRandom(): T {
        return getRandomItemFromArray(this.unlocked) ?? null;
    }
}

export class Level {
    constructor(protected _value: number, readonly min: number, readonly max: number) {}

    set value(setTo: number) {
        if (setTo >= this.max) {
            this._value = this.max;
        }
        else if (setTo <= this.min) {
            this._value = this.min;
        }
        else {
            this._value = setTo;
        }
    }

    get value() {
        return this._value;
    }

    isMinLevel(): boolean {
        return this.value <= this.min;
    }

    isMaxLevel(): boolean {
        return this.value >= this.max;
    }

    setMaxLevel(): void {
        this.value = this.max;
    }

    addValue(change: number): number {
        this.value = this.value + change;
        return this.value;
    }

    increment(): number {
        return this.addValue(1);
    }

    decrement(): number {
        return this.addValue(-1);
    }
}