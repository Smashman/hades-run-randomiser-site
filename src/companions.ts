import { getRandomItemFromArray, Item, Options } from './utils';
import { ItemList } from './utils';

export class Companion extends Item {
    level: number = 1;
    isMaxCodex: boolean = false;
    constructor(public name: string, public giver: string) {
        super();
    }

    static maxLevel: number = 5;

    get isMaxLevel(): boolean {
        return this.level === Companion.maxLevel;
    }
}

export class Companions extends ItemList<Companion> {
    getRandomCompanion(companionOptions?: CompanionOptions): Companion | null {
        const companions = (companionOptions?.ignoreMaxCodex && !this.isAllUnlockedMaxLevel) ? [...this.unlocked.filter(companion => !companion.isMaxCodex)] : [...this.unlocked];
        return getRandomItemFromArray(companions) ?? null;
    }

    maxCodexAll() {
        this.items.forEach(companion => companion.isMaxCodex = true);
    }

    get isAllMaxCodex() {
        return this.items.every(companion => companion.isMaxCodex);
    }

    maxLevelAll() {
        this.items.forEach(companion => companion.level = Companion.maxLevel);
    }

    get isAllUnlockedMaxLevel() {
        return this.unlocked.every(companion => companion.isMaxLevel);
    }

    get isAllMaxLevel() {
        return this.items.every(companion => companion.isMaxLevel);
    }
}

export interface CompanionOptions extends Options {
    ignoreMaxCodex?: boolean;
}