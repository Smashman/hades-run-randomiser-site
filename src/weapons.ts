import { Item, ItemList, Options, getRandomItemFromArray, Level, StorableItemListData, StorableItemData, Resource } from './utils';

export interface WeaponOptions extends Options {
    randomAspect?: boolean;
}

export type StorableWeaponsData = StorableItemListData<StorableWeaponData>;

export class Weapons extends ItemList<Weapon> {
    unlockAll() {
        super.unlockAll();
        this.items.forEach(weapon => weapon.unlockAllAspects());
    }

    maxLevelAll() {
        this.items.forEach(weapon => weapon.maxLevelAllAspects());
    }

    get resourceTotal() {
        return this.items.reduce((total, weapon) => total + weapon.resourceTotal, 0);
    }

    get resourceSpent() {
        return this.items.reduce((total, weapon) => total + weapon.resourceSpent, 0);
    }

    get resourceRequired() {
        return this.resourceTotal - this.resourceSpent;
    }
}

interface StorableWeaponData extends StorableItemData {
    aspects: StorableWeaponAspectData[];
}

export class Weapon extends Item {
    aspects: WeaponAspect[] = [];
    canBeLocked: boolean = true;

    constructor(public name: string, public title: string, public shortName: string, public isUnlocked: boolean = false){
        super();
        this.canBeLocked = !isUnlocked;
    }

    get unlockedAspects() {
        return this.aspects.filter(aspect => aspect.isUnlocked);
    }

    lock() {
        if (this.canBeLocked) {
            super.lock();
        }
    }

    addAspect(aspect: WeaponAspect) {
        this.aspects = [...this.aspects, aspect];
        return this;
    }

    addAspects(aspects: WeaponAspect[]) {
        aspects.forEach(aspect => this.addAspect(aspect));
        return this;
    }

    unlockAllAspects() {
        this.aspects.forEach(aspect => aspect.unlock());
    }

    maxLevelAllAspects() {
        this.aspects.forEach(aspect => aspect.level.setMaxLevel());
    }

    getRandomAspect(): WeaponAspect {
        return getRandomItemFromArray(this.unlockedAspects) ?? null;
    }

    get resourceTotal() {
        return this.aspects.reduce((total, aspect) => total + aspect.resource.total, 0);
    }

    get resourceSpent() {
        return this.aspects.reduce((total, aspect) => total + aspect.resource.spent, 0);
    }

    get resourceRequired() {
        return this.resourceTotal - this.resourceSpent;
    }

    toStorableData(): StorableWeaponData {
        const itemData = super.toStorableData();
        return {
            ...itemData,
            aspects: this.aspects.map(aspect => aspect.toStorableData()),
        }
    }

    fromStoredData(storedData: StorableWeaponData) {
        super.fromStoredData(storedData);
        storedData.aspects.map((data, index) => {
            this.aspects[index].fromStoredData(data);
        });
    }
}

interface StorableWeaponAspectData extends StorableItemData {
    level: number;
}

export class WeaponAspect extends Item {
    level: Level;
    resource: Resource;
    maxLevel: number;
    canBeLocked: boolean = true;

    constructor(public name: string, public icon: string, private readonly levelCosts: number[], public isHidden: boolean = false, public isUnlocked: boolean = false){
        super();
        const onLevelChange = this.onLevelChange.bind(this);
        this.maxLevel = levelCosts.length;
        this.level = new Level(0, WeaponAspect.minLevel, this.maxLevel, onLevelChange);
        this.resource = new Resource(levelCosts, this.level, 'blood');
        this.canBeLocked = !isUnlocked;
    }

    static minLevel = 0;

    lock() {
        if (this.canBeLocked) {
            super.lock();
        }
    }

    onLevelChange(level: number) {
        this.isUnlocked = this.canBeLocked ? level > WeaponAspect.minLevel : true;
    }

    toStorableData(): StorableWeaponAspectData {
        const itemData = super.toStorableData();
        return {
            ...itemData,
            level: this.level.value,
        }
    }

    fromStoredData(storableData: StorableWeaponAspectData) {
        super.fromStoredData(storableData);
        this.level.value = storableData.level;
    }
}