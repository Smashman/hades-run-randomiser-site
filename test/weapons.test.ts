import { Weapon, WeaponAspect, Weapons } from '../src/weapons';
import * as utils from '../src/utils';
import { getRandomItemFromArrayMock, fakeIcon } from './helpers';

describe('WeaponAspect class', () => {
    it('should create expected WeaponAspect class instance', () => {
        const weaponAspect = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);

        expect(weaponAspect).toMatchSnapshot();
    });

    it('should create expected WeaponAspect class instance with more parameters', () => {
        const weaponAspect = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1], true, true);

        expect(weaponAspect).toMatchSnapshot();
    });

    it('should return false for isUnlocked after lock is called and canBeLocked is true', () => {
        const weaponAspect = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);

        expect(weaponAspect.canBeLocked).toBe(true);

        weaponAspect.unlock();
        expect(weaponAspect.isUnlocked).toBe(true);

        weaponAspect.lock();
        expect(weaponAspect.isUnlocked).toBe(false);
    });

    it('should return false for isUnlocked after lock is called and canBeLocked is false', () => {
        const weaponAspect = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1], false, true);

        expect(weaponAspect.canBeLocked).toBe(false);
        expect(weaponAspect.isUnlocked).toBe(true);

        weaponAspect.lock();
        expect(weaponAspect.isUnlocked).toBe(true);
    });

    it('should set isUnlocked to true when changing level above 0 when canBeLocked is true', () => {
        const weaponAspect = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);

        expect(weaponAspect.canBeLocked).toBe(true);
        expect(weaponAspect.isUnlocked).toBe(false);

        weaponAspect.level.value = 1;

        expect(weaponAspect.isUnlocked).toBe(true);
    });

    it('should set isUnlocked to false when changing level to 0 when canBeLocked is true', () => {
        const weaponAspect = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);

        expect(weaponAspect.canBeLocked).toBe(true);
        expect(weaponAspect.isUnlocked).toBe(false);

        weaponAspect.level.value = 1;

        expect(weaponAspect.isUnlocked).toBe(true);

        weaponAspect.level.value = 0;

        expect(weaponAspect.isUnlocked).toBe(false);
    });

    it('should not change isUnlocked when changing level above 0 when canBeLocked is false', () => {
        const weaponAspect = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1], false, true);

        expect(weaponAspect.canBeLocked).toBe(false);
        expect(weaponAspect.isUnlocked).toBe(true);

        weaponAspect.level.value = 1;

        expect(weaponAspect.isUnlocked).toBe(true);
    });

    it('should not change isUnlocked when changing level to 0 when canBeLocked is false', () => {
        const weaponAspect = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1], false, true);

        expect(weaponAspect.canBeLocked).toBe(false);
        expect(weaponAspect.isUnlocked).toBe(true);

        weaponAspect.level.value = 1;

        expect(weaponAspect.isUnlocked).toBe(true);

        weaponAspect.level.value = 0;

        expect(weaponAspect.isUnlocked).toBe(true);
    });

    it('should change resource values when level is updated', () => {
        const weaponAspect = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);

        expect(weaponAspect.resource.total).toBe(5);
        expect(weaponAspect.resource.spent).toBe(0);
        expect(weaponAspect.resource.required).toBe(5);

        weaponAspect.level.value = 4;

        expect(weaponAspect.resource.total).toBe(5);
        expect(weaponAspect.resource.spent).toBe(4);
        expect(weaponAspect.resource.required).toBe(1);
    });

    describe('toStorableData', () => {
        it('should return expected storable data', () => {
            const weaponAspect = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);
    
            expect(weaponAspect.toStorableData()).toMatchSnapshot();
        });

        it('should return expected storable data with level', () => {
            const weaponAspect = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);
            weaponAspect.level.value = 5;
    
            expect(weaponAspect.toStorableData()).toMatchSnapshot();
        });
    });

    describe('fromStoredData', () => {
        it('should retun expected Weapon instance after loading from data', () => {
            const weaponAspect = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);
            weaponAspect.fromStoredData({
                isUnlocked: true,
                level: 2,
            });

            expect(weaponAspect).toMatchSnapshot();
        });
    });
});

// Skybreaker is a legendary weapon, a 'Mistblade', in my D&D campaign.
// Tarak is a player's character who now owns it.
// Shokrug is its previous owner.
// It fit too well not to use...

describe('Weapon class', () => {
    it('should create expected Weapon class instance', () => {
        const weapon = new Weapon('Skybreaker', 'Mistblade', 'sword');

        expect(weapon).toMatchSnapshot();
    });

    it('should add aspect to Weapon', () => {
        const weaponAspect = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);
        const weapon = new Weapon('Skybreaker', 'Mistblade', 'sword').addAspect(weaponAspect);

        expect(weapon.aspects).toHaveLength(1);
        expect(weapon.aspects).toContain(weaponAspect);
    });

    it('should add aspects to Weapon', () => {
        const weaponAspect1 = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);
        const weaponAspect2 = new WeaponAspect('Shokrug', fakeIcon, [1, 2, 3, 4, 5]);
        const weapon = new Weapon('Skybreaker', 'Mistblade', 'sword').addAspects([weaponAspect1, weaponAspect2]);

        expect(weapon.aspects).toHaveLength(2);
        expect(weapon.aspects).toContain(weaponAspect1);
        expect(weapon.aspects).toContain(weaponAspect2);
    });

    it('should return no items for unlockedAspects if no aspects are unlocked', () => {
        const weaponAspect1 = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);
        const weaponAspect2 = new WeaponAspect('Shokrug', fakeIcon, [1, 2, 3, 4, 5]);
        const weapon = new Weapon('Skybreaker', 'Mistblade', 'sword').addAspects([weaponAspect1, weaponAspect2]);

        expect(weapon.unlockedAspects).toHaveLength(0);
    });

    it('should return expected items for unlockedAspects when a subset of aspects is unlocked', () => {
        const weaponAspect1 = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);
        weaponAspect1.unlock();

        const weaponAspect2 = new WeaponAspect('Shokrug', fakeIcon, [1, 2, 3, 4, 5]);

        const weapon = new Weapon('Skybreaker', 'Mistblade', 'sword').addAspects([weaponAspect1, weaponAspect2]);

        expect(weapon.unlockedAspects).toHaveLength(1);
        expect(weapon.unlockedAspects).toContain(weaponAspect1);
        expect(weapon.unlockedAspects).not.toContain(weaponAspect2);
    });

    it('should return all items for unlockedAspects when all aspects are unlocked', () => {
        const weaponAspect1 = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);
        weaponAspect1.unlock();

        const weaponAspect2 = new WeaponAspect('Shokrug', fakeIcon, [1, 2, 3, 4, 5]);
        weaponAspect2.unlock();

        const weapon = new Weapon('Skybreaker', 'Mistblade', 'sword').addAspects([weaponAspect1, weaponAspect2]);

        expect(weapon.unlockedAspects).toHaveLength(2);
        expect(weapon.unlockedAspects).toContain(weaponAspect1);
        expect(weapon.unlockedAspects).toContain(weaponAspect2);
    });

    it('should set all aspects to unlocked when calling unlockAllAspects', () => {
        const weaponAspect1 = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);
        const weaponAspect2 = new WeaponAspect('Shokrug', fakeIcon, [1, 2, 3, 4, 5]);
        const weapon = new Weapon('Skybreaker', 'Mistblade', 'sword').addAspects([weaponAspect1, weaponAspect2]);

        weapon.unlockAllAspects();

        expect(weapon.unlockedAspects).toHaveLength(2);
        expect(weapon.unlockedAspects).toContain(weaponAspect1);
        expect(weapon.unlockedAspects).toContain(weaponAspect2);
    });

    it('should set all aspects to max level when calling maxLevelAllAspects', () => {
        const weaponAspect1 = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);
        const weaponAspect2 = new WeaponAspect('Shokrug', fakeIcon, [1, 2, 3, 4, 5]);
        const weapon = new Weapon('Skybreaker', 'Mistblade', 'sword').addAspects([weaponAspect1, weaponAspect2]);

        weapon.maxLevelAllAspects();

        expect(weaponAspect1.level.value).toBe(5);
        expect(weaponAspect2.level.value).toBe(5);
    });

    it('should return false for isUnlocked after lock is called and canBeLocked is true', () => {
        const weaponAspect = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);
        const weapon = new Weapon('Skybreaker', 'Mistblade', 'sword').addAspect(weaponAspect);

        expect(weapon.canBeLocked).toBe(true);

        weapon.unlock();
        expect(weapon.isUnlocked).toBe(true);

        weapon.lock();
        expect(weapon.isUnlocked).toBe(false);
    });

    it('should return false for isUnlocked after lock is called and canBeLocked is false', () => {
        const weaponAspect = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);
        const weapon = new Weapon('Skybreaker', 'Mistblade', 'sword', true).addAspect(weaponAspect);

        expect(weapon.canBeLocked).toBe(false);
        expect(weapon.isUnlocked).toBe(true);

        weapon.lock();
        expect(weapon.isUnlocked).toBe(true);
    });

    describe('getRandomAspect', () => {
        let getRandomItemFromArraySpy: jest.SpyInstance;

        let weaponAspect1: WeaponAspect;
        let weaponAspect2: WeaponAspect;

        let weapon: Weapon;

        beforeEach(() => {
            weaponAspect1 = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);
            weaponAspect2 = new WeaponAspect('Shokrug', fakeIcon, [1, 2, 3, 4, 5]);
            weapon = new Weapon('Skybreaker', 'Mistblade', 'sword').addAspects([weaponAspect1, weaponAspect2]);

            getRandomItemFromArraySpy = jest.spyOn(utils, 'getRandomItemFromArray').mockImplementation(getRandomItemFromArrayMock);
        });

        afterEach(() => {
            jest.spyOn(utils, 'getRandomItemFromArray').mockRestore();
        });

        it('should return null if no aspects are unlocked', () => {
            const randomAspect = weapon.getRandomAspect();

            expect(randomAspect).toBe(null);
        });

        it('should return expected aspect if all aspect are unlocked', () => {
            weapon.unlockAllAspects();

            const randomAspect = weapon.getRandomAspect();

            expect(getRandomItemFromArraySpy).toBeCalledWith([weaponAspect1, weaponAspect2]);
            expect(randomAspect).toBe(weaponAspect2);
        });

        it('should return expected companion when a subset of companions are unlocked', () => {
            weaponAspect1.unlock();

            const randomCompanion = weapon.getRandomAspect();

            expect(getRandomItemFromArraySpy).toBeCalledWith([weaponAspect1]);
            expect(randomCompanion).toBe(weaponAspect1);
        });
    });

    describe('Resources', () => {
        let weaponAspect1: WeaponAspect;
        let weaponAspect2: WeaponAspect;

        let weapon: Weapon;

        beforeEach(() => {
            weaponAspect1 = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);
            weaponAspect2 = new WeaponAspect('Shokrug', fakeIcon, [1, 2, 3, 4, 5]);
            weapon = new Weapon('Skybreaker', 'Mistblade', 'sword').addAspects([weaponAspect1, weaponAspect2]);
        });

        it('should calculate resource values from all aspects', () => {
            expect(weapon.resourceTotal).toBe(20);
            expect(weapon.resourceSpent).toBe(0);
            expect(weapon.resourceRequired).toBe(20);
        });

        it('should calculate resource values from all aspects when aspects are leveled', () => {
            weaponAspect1.level.value = 2;
            weaponAspect2.level.value = 3;

            expect(weapon.resourceTotal).toBe(20);
            expect(weapon.resourceSpent).toBe(8);
            expect(weapon.resourceRequired).toBe(12);
        });
    });

    describe('toStorableData', () => {
        it('should return expected storable data', () => {
            const weaponAspect1 = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);
            const weaponAspect2 = new WeaponAspect('Shokrug', fakeIcon, [1, 2, 3, 4, 5]);
            const weapon = new Weapon('Skybreaker', 'Mistblade', 'sword').addAspects([weaponAspect1, weaponAspect2]);
    
            expect(weapon.toStorableData()).toMatchSnapshot();
        });

        it('should return expected storable data with aspects with level', () => {
            const weaponAspect1 = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);
            const weaponAspect2 = new WeaponAspect('Shokrug', fakeIcon, [1, 2, 3, 4, 5]);
            const weapon = new Weapon('Skybreaker', 'Mistblade', 'sword').addAspects([weaponAspect1, weaponAspect2]);
            weapon.unlock();
            weaponAspect1.level.value = 2;
            weaponAspect2.level.value = 4;
    
            expect(weapon.toStorableData()).toMatchSnapshot();
        });
    });

    describe('fromStoredData', () => {
        it('should retun expected Weapons instance after loading from data', () => {
            const weaponAspect1 = new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]);
            const weaponAspect2 = new WeaponAspect('Shokrug', fakeIcon, [1, 2, 3, 4, 5]);
            const weapon = new Weapon('Skybreaker', 'Mistblade', 'sword').addAspects([weaponAspect1, weaponAspect2]);
            weapon.fromStoredData({
                isUnlocked: true,
                aspects: [
                    {
                        isUnlocked: true,
                        level: 3,
                    },
                    {
                        isUnlocked: false,
                        level: 0,
                    }
                ]
            });

            expect(weapon).toMatchSnapshot();
        });
    });
});

describe('Weapons class', () => {
    it('should unlock all aspects as well as all weapons when calling unlockAll', () => {
        const weapon1 = new Weapon('Skybreaker', 'Mistblade', 'sword').addAspects([new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]), new WeaponAspect('Shokrug', fakeIcon, [1, 2, 3, 4, 5])]);
        const weapon2 = new Weapon('Lightsaber', 'Weapon of the Jedi', 'laser sword').addAspects([new WeaponAspect('Rey', fakeIcon, [1, 1, 1, 1, 1]), new WeaponAspect('Luke', fakeIcon, [1, 2, 3, 4, 5]), new WeaponAspect('Anakin', fakeIcon, [3, 3, 3, 3, 3])]);
        const weapons = new Weapons().addItems([weapon1, weapon2]);

        expect(weapons.unlocked).toHaveLength(0);
        expect(weapon1.unlockedAspects).toHaveLength(0);
        expect(weapon2.unlockedAspects).toHaveLength(0);
        expect(weapon1.isUnlocked).toBe(false);
        expect(weapon2.isUnlocked).toBe(false);

        weapons.unlockAll();

        expect(weapons.unlocked).toHaveLength(2);
        expect(weapon1.unlockedAspects).toHaveLength(2);
        expect(weapon2.unlockedAspects).toHaveLength(3);
        expect(weapon1.isUnlocked).toBe(true);
        expect(weapon2.isUnlocked).toBe(true);
    });

    it('should max level all aspects of all weapons when calling maxLevelAll', () => {
        const weapon1 = new Weapon('Skybreaker', 'Mistblade', 'sword').addAspects([new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]), new WeaponAspect('Shokrug', fakeIcon, [1, 2, 3, 4, 5])]);
        const weapon2 = new Weapon('Lightsaber', 'Weapon of the Jedi', 'laser sword').addAspects([new WeaponAspect('Rey', fakeIcon, [1, 1, 1, 1, 1]), new WeaponAspect('Luke', fakeIcon, [1, 2, 3, 4, 5]), new WeaponAspect('Anakin', fakeIcon, [3, 3, 3, 3, 3])]);
        const weapons = new Weapons().addItems([weapon1, weapon2]);

        expect(weapon1.aspects[0].level.value).toBe(0);
        expect(weapon1.aspects[1].level.value).toBe(0);
        expect(weapon2.aspects[0].level.value).toBe(0);
        expect(weapon2.aspects[1].level.value).toBe(0);
        expect(weapon2.aspects[2].level.value).toBe(0);

        weapons.maxLevelAll();

        expect(weapon1.aspects[0].level.value).toBe(5);
        expect(weapon1.aspects[1].level.value).toBe(5);
        expect(weapon2.aspects[0].level.value).toBe(5);
        expect(weapon2.aspects[1].level.value).toBe(5);
        expect(weapon2.aspects[2].level.value).toBe(5);
    });

    it('should calculate resource values from all weapons', () => {
        const weapon1 = new Weapon('Skybreaker', 'Mistblade', 'sword').addAspects([new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]), new WeaponAspect('Shokrug', fakeIcon, [1, 2, 3, 4, 5])]);
        const weapon2 = new Weapon('Lightsaber', 'Weapon of the Jedi', 'laser sword').addAspects([new WeaponAspect('Rey', fakeIcon, [1, 1, 1, 1, 1]), new WeaponAspect('Luke', fakeIcon, [1, 2, 3, 4, 5]), new WeaponAspect('Anakin', fakeIcon, [3, 3, 3, 3, 3])]);
        const weapons = new Weapons().addItems([weapon1, weapon2]);

        expect(weapons.resourceTotal).toBe(55);
        expect(weapons.resourceSpent).toBe(0);
        expect(weapons.resourceRequired).toBe(55);
    });

    it('should calculate resource values from all weapons when aspects are leveled', () => {
        const weapon1 = new Weapon('Skybreaker', 'Mistblade', 'sword').addAspects([new WeaponAspect('Tarak', fakeIcon, [1, 1, 1, 1, 1]), new WeaponAspect('Shokrug', fakeIcon, [1, 2, 3, 4, 5])]);
        const weapon2 = new Weapon('Lightsaber', 'Weapon of the Jedi', 'laser sword').addAspects([new WeaponAspect('Rey', fakeIcon, [1, 1, 1, 1, 1]), new WeaponAspect('Luke', fakeIcon, [1, 2, 3, 4, 5]), new WeaponAspect('Anakin', fakeIcon, [3, 3, 3, 3, 3])]);
        const weapons = new Weapons().addItems([weapon1, weapon2]);

        weapon1.aspects[0].level.value = 2;
        weapon1.aspects[1].level.value = 4;
        weapon2.aspects[0].level.value = 3;
        weapon2.aspects[1].level.value = 0;
        weapon2.aspects[2].level.value = 5;

        expect(weapons.resourceTotal).toBe(55);
        expect(weapons.resourceSpent).toBe(30);
        expect(weapons.resourceRequired).toBe(25);
    });
});