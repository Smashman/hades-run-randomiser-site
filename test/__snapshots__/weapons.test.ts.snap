// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weapon class fromStoredData should retun expected Weapons instance after loading from data 1`] = `
Weapon {
  "aspects": Array [
    WeaponAspect {
      "canBeLocked": true,
      "icon": "base64:icon!",
      "isHidden": false,
      "isUnlocked": true,
      "level": Level {
        "_value": 3,
        "max": 5,
        "min": 0,
        "onChange": [Function],
      },
      "levelCosts": Array [
        1,
        1,
        1,
        1,
        1,
      ],
      "maxLevel": 5,
      "name": "Tarak",
      "resource": Resource {
        "costs": Array [
          1,
          1,
          1,
          1,
          1,
        ],
        "level": Level {
          "_value": 3,
          "max": 5,
          "min": 0,
          "onChange": [Function],
        },
        "name": "blood",
        "total": 5,
      },
    },
    WeaponAspect {
      "canBeLocked": true,
      "icon": "base64:icon!",
      "isHidden": false,
      "isUnlocked": false,
      "level": Level {
        "_value": 0,
        "max": 5,
        "min": 0,
        "onChange": [Function],
      },
      "levelCosts": Array [
        1,
        2,
        3,
        4,
        5,
      ],
      "maxLevel": 5,
      "name": "Shokrug",
      "resource": Resource {
        "costs": Array [
          1,
          2,
          3,
          4,
          5,
        ],
        "level": Level {
          "_value": 0,
          "max": 5,
          "min": 0,
          "onChange": [Function],
        },
        "name": "blood",
        "total": 15,
      },
    },
  ],
  "canBeLocked": true,
  "isUnlocked": true,
  "name": "Skybreaker",
  "shortName": "sword",
  "title": "Mistblade",
}
`;

exports[`Weapon class should create expected Weapon class instance 1`] = `
Weapon {
  "aspects": Array [],
  "canBeLocked": true,
  "isUnlocked": false,
  "name": "Skybreaker",
  "shortName": "sword",
  "title": "Mistblade",
}
`;

exports[`Weapon class toStorableData should return expected storable data 1`] = `
Object {
  "aspects": Array [
    Object {
      "isUnlocked": false,
      "level": 0,
    },
    Object {
      "isUnlocked": false,
      "level": 0,
    },
  ],
  "isUnlocked": false,
}
`;

exports[`Weapon class toStorableData should return expected storable data with aspects with level 1`] = `
Object {
  "aspects": Array [
    Object {
      "isUnlocked": true,
      "level": 2,
    },
    Object {
      "isUnlocked": true,
      "level": 4,
    },
  ],
  "isUnlocked": true,
}
`;

exports[`WeaponAspect class fromStoredData should retun expected Weapon instance after loading from data 1`] = `
WeaponAspect {
  "canBeLocked": true,
  "icon": "base64:icon!",
  "isHidden": false,
  "isUnlocked": true,
  "level": Level {
    "_value": 2,
    "max": 5,
    "min": 0,
    "onChange": [Function],
  },
  "levelCosts": Array [
    1,
    1,
    1,
    1,
    1,
  ],
  "maxLevel": 5,
  "name": "Tarak",
  "resource": Resource {
    "costs": Array [
      1,
      1,
      1,
      1,
      1,
    ],
    "level": Level {
      "_value": 2,
      "max": 5,
      "min": 0,
      "onChange": [Function],
    },
    "name": "blood",
    "total": 5,
  },
}
`;

exports[`WeaponAspect class should create expected WeaponAspect class instance 1`] = `
WeaponAspect {
  "canBeLocked": true,
  "icon": "base64:icon!",
  "isHidden": false,
  "isUnlocked": false,
  "level": Level {
    "_value": 0,
    "max": 5,
    "min": 0,
    "onChange": [Function],
  },
  "levelCosts": Array [
    1,
    1,
    1,
    1,
    1,
  ],
  "maxLevel": 5,
  "name": "Tarak",
  "resource": Resource {
    "costs": Array [
      1,
      1,
      1,
      1,
      1,
    ],
    "level": Level {
      "_value": 0,
      "max": 5,
      "min": 0,
      "onChange": [Function],
    },
    "name": "blood",
    "total": 5,
  },
}
`;

exports[`WeaponAspect class should create expected WeaponAspect class instance with more parameters 1`] = `
WeaponAspect {
  "canBeLocked": false,
  "icon": "base64:icon!",
  "isHidden": true,
  "isUnlocked": true,
  "level": Level {
    "_value": 0,
    "max": 5,
    "min": 0,
    "onChange": [Function],
  },
  "levelCosts": Array [
    1,
    1,
    1,
    1,
    1,
  ],
  "maxLevel": 5,
  "name": "Tarak",
  "resource": Resource {
    "costs": Array [
      1,
      1,
      1,
      1,
      1,
    ],
    "level": Level {
      "_value": 0,
      "max": 5,
      "min": 0,
      "onChange": [Function],
    },
    "name": "blood",
    "total": 5,
  },
}
`;

exports[`WeaponAspect class toStorableData should return expected storable data 1`] = `
Object {
  "isUnlocked": false,
  "level": 0,
}
`;

exports[`WeaponAspect class toStorableData should return expected storable data with level 1`] = `
Object {
  "isUnlocked": true,
  "level": 5,
}
`;
